File: python/protein-translation/protein_translation.py
def proteins(strand):
    proteins = {
        "Methionine": ["AUG"],
        "Phenylalanine": ["UUU", "UUC"],
        "Leucine": ["UUA", "UUG"],
        "Serine": ["UCU", "UCC", "UCA", "UCG"],
        "Tyrosine": ["UAU", "UAC"],
        "Cysteine": ["UGU", "UGC"],
        "Tryptophan": ["UGG"],
        "STOP": ["UAA", "UAG", "UGA"],
    }

    n = 3
    rna_list = [strand[i : i + n] for i in range(0, len(strand), n)]
    protein_list = []

    for rna in rna_list:
        if rna in proteins["STOP"]:
            break
        for key, values in proteins.items():
            for value in values:
                if rna == value:
                    protein_list.append(key)

    return protein_list

File: javascript/protein-translation/protein-translation.js

export const translate = (rnaSequence = '') => {
  let codonMap = {
    AUG:'Methionine',
    UUU:'Phenylalanine',
    UUC:'Phenylalanine',
    UUA:'Leucine',
    UUG:'Leucine',
    UCU:'Serine',
    UCC:'Serine',
    UCA:'Serine',
    UCG:'Serine',
    UAU:'Tyrosine',
    UAC:'Tyrosine',
    UGU:'Cysteine',
    UGC:'Cysteine',
    UGG:'Tryptophan',
    UAA:'STOP',
    UAG:'STOP',
    UGA:'STOP',
  };

  let proteins = [];

  for (let i = 0; i < rnaSequence.length; i += 3) {
    let codon = rnaSequence.slice(i, i + 3);
    
    if (!(codon in codonMap)) {
      throw Error('Invalid codon');
    }

    if (codonMap[codon] === 'STOP') {
      break;
    }

    proteins.push(codonMap[codon]);
  }

  return proteins;
};
